### PR TITLE
Use urlib3 with certifi for downloading files over http and https

### PR DESCRIPTION
**Pull request by doublethefish**
Because urlib3 doesn't support FTP, we use urlretrieve for non-http downloads.

This fixes cert errors like the following:
  `OSError: [Errno socket error] [Errno socket error] [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1123)`

The above error appears to be caused by the certs not being found,
whether in a venv or not. Using ceritfi explicitly seems to fix that.

Tested with:
```
#!/bin/bash
# exit when any command fails
set -e
set -o pipefail

if [ ! -d py3env/bin/ ]; then
  python3 -m virtualenv -p python3 py3env
fi

source py3env/bin/activate

brew install autoconf automake libtool libffi pkg-config 2>&1 | sed -e 's/^/BREW: /g'
brew link libtool 2>&1 | sed -e 's/^/BREW LINK: /g'
pip install -e "$KIVY_IOS_CHECKOUT" 2>&1 | sed -e 's/^/PIP INSTALL kivy: /g'
python3 -m pip install --upgrade certifi 2>&1 | sed -e 's/^/PIP CERTIFY: /g'

# fails at this stage with the above OSError without the patch
toolchain build kivy 2>&1 | sed -e 's/^/build kivy: /g'
```

### CODE REVIEW

**Comment by misl6** [comment]
Hi, can you please be more verbose about your current environment? (MacOS version, Python version)

The CI builds are not affected by this issue, so I would like to better dig into the issue before reviewing your changes.

**Comment by doublethefish** [comment]
Sure thing. If the following is not enough information on the env, let me know (and reciprocate verbosity ;)

Running the bash above should get you into the right state.

# General
macOS 11.2.1 Bug Sur (also tried Yosemite)
PYTHON: Python 3.8.3
virtualenv 20.4.2

# Slightly more verbose python info from pylint
PYLINT: Python 3.8.3 (v3.8.3:6f8c8320e9, May 13 2020, 16:29:34)
PYLINT: [Clang 6.0 (clang-600.0.57)]

# Pip from venv
PIP FREEZE: arrow==1.0.0
PIP FREEZE: astroid==2.5
PIP FREEZE: binaryornot==0.4.4
PIP FREEZE: certifi==2020.12.5
PIP FREEZE: chardet==4.0.0
PIP FREEZE: click==7.1.2
PIP FREEZE: cookiecutter==1.7.2
PIP FREEZE: docopt==0.6.2
PIP FREEZE: idna==2.10
PIP FREEZE: isort==5.7.0
PIP FREEZE: Jinja2==2.11.3
PIP FREEZE: jinja2-time==0.2.0
PIP FREEZE: -e git+git@github.com:kivy/kivy-ios.git@2cb084c26b7e9b45b027961cded13521b8db5b6b#egg=kivy_ios
PIP FREEZE: lazy-object-proxy==1.5.2
PIP FREEZE: MarkupSafe==1.1.1
PIP FREEZE: mccabe==0.6.1
PIP FREEZE: openstep-parser==1.5.3
PIP FREEZE: pbxproj==3.2.1
PIP FREEZE: Pillow==8.1.0
PIP FREEZE: poyo==0.5.0
PIP FREEZE: pylint==2.7.1
PIP FREEZE: python-dateutil==2.8.1
PIP FREEZE: python-slugify==4.0.1
PIP FREEZE: requests==2.25.1
PIP FREEZE: sh==1.14.1
PIP FREEZE: six==1.15.0
PIP FREEZE: text-unidecode==1.3
PIP FREEZE: toml==0.10.2
PIP FREEZE: urllib3==1.26.3
PIP FREEZE: wrapt==1.12.1

# Misc relevant homebrew info
BREW VER: Homebrew 3.0.2-102-g25d0987
BREW VER: Homebrew/homebrew-core (git revision a6520a; last commit 2021-02-27)
BREW VER: Homebrew/homebrew-cask (git revision e8a67; last commit 2021-02-27)
BREW INFO autoconf: autoconf: stable 2.69 (bottled)
BREW INFO autoconf: Automatic configure script builder
BREW INFO autoconf: https://www.gnu.org/software/autoconf
BREW INFO autoconf: /usr/local/Cellar/autoconf/2.69 (71 files, 3.0MB) *
BREW INFO autoconf:   Poured from bottle on 2019-09-22 at 23:08:58
BREW INFO autoconf: From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/autoconf.rb
BREW INFO autoconf: License: GPL-2.0-or-later
BREW INFO autoconf: ==> Caveats
BREW INFO autoconf: Emacs Lisp files have been installed to:
BREW INFO autoconf:   /usr/local/share/emacs/site-lisp/autoconf
BREW INFO autoconf: ==> Analytics
BREW INFO autoconf: install: 113,109 (30 days), 319,758 (90 days), 1,170,529 (365 days)
BREW INFO autoconf: install-on-request: 18,338 (30 days), 53,147 (90 days), 221,120 (365 days)
BREW INFO autoconf: build-error: 0 (30 days)
BREW INFO automake: automake: stable 1.16.3 (bottled)
BREW INFO automake: Tool for generating GNU Standards-compliant Makefiles
BREW INFO automake: https://www.gnu.org/software/automake/
BREW INFO automake: /usr/local/Cellar/automake/1.16.3 (131 files, 3.4MB) *
BREW INFO automake:   Poured from bottle on 2021-02-23 at 19:50:44
BREW INFO automake: From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/automake.rb
BREW INFO automake: License: GPL-2.0-or-later
BREW INFO automake: ==> Dependencies
BREW INFO automake: Required: autoconf
BREW INFO automake: ==> Analytics
BREW INFO automake: install: 68,281 (30 days), 229,161 (90 days), 947,251 (365 days)
BREW INFO automake: install-on-request: 44,914 (30 days), 146,730 (90 days), 603,349 (365 days)
BREW INFO automake: build-error: 0 (30 days)
BREW INFO libtool: libtool: stable 2.4.6 (bottled)
BREW INFO libtool: Generic library support script
BREW INFO libtool: https://www.gnu.org/software/libtool/
BREW INFO libtool: /usr/local/Cellar/libtool/2.4.6_2 (71 files, 3.7MB) *
BREW INFO libtool:   Poured from bottle on 2020-06-25 at 03:00:19
BREW INFO libtool: From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/libtool.rb
BREW INFO libtool: License: GPL-2.0-or-later
BREW INFO libtool: ==> Caveats
BREW INFO libtool: In order to prevent conflicts with Apple's own libtool we have prepended a "g"
BREW INFO libtool: so, you have instead: glibtool and glibtoolize.
BREW INFO libtool: ==> Analytics
BREW INFO libtool: install: 197,439 (30 days), 583,063 (90 days), 2,199,784 (365 days)
BREW INFO libtool: install-on-request: 27,646 (30 days), 85,938 (90 days), 402,173 (365 days)
BREW INFO libtool: build-error: 0 (30 days)
BREW INFO libffi: libffi: stable 3.3 (bottled), HEAD [keg-only]
BREW INFO libffi: Portable Foreign Function Interface library
BREW INFO libffi: https://sourceware.org/libffi/
BREW INFO libffi: /usr/local/Cellar/libffi/3.3_2 (17 files, 540.5KB)
BREW INFO libffi:   Poured from bottle on 2021-01-13 at 03:00:19
BREW INFO libffi: From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/libffi.rb
BREW INFO libffi: License: MIT
BREW INFO libffi: ==> Options
BREW INFO libffi: --HEAD
BREW INFO libffi: 	Install HEAD version
BREW INFO libffi: ==> Caveats
BREW INFO libffi: libffi is keg-only, which means it was not symlinked into /usr/local,
BREW INFO libffi: because macOS already provides this software and installing another version in
BREW INFO libffi: parallel can cause all kinds of trouble.
BREW INFO libffi:
BREW INFO libffi: For compilers to find libffi you may need to set:
BREW INFO libffi:   export LDFLAGS="-L/usr/local/opt/libffi/lib"
BREW INFO libffi:   export CPPFLAGS="-I/usr/local/opt/libffi/include"
BREW INFO libffi:
BREW INFO libffi: For pkg-config to find libffi you may need to set:
BREW INFO libffi:   export PKG_CONFIG_PATH="/usr/local/opt/libffi/lib/pkgconfig"
BREW INFO libffi:
BREW INFO libffi: ==> Analytics
BREW INFO libffi: install: 344,860 (30 days), 920,944 (90 days), 2,884,584 (365 days)
BREW INFO libffi: install-on-request: 17,398 (30 days), 50,011 (90 days), 128,207 (365 days)
BREW INFO libffi: build-error: 0 (30 days)
BREW INFO pkg-config: pkg-config: stable 0.29.2 (bottled)
BREW INFO pkg-config: Manage compile and link flags for libraries
BREW INFO pkg-config: https://freedesktop.org/wiki/Software/pkg-config/
BREW INFO pkg-config: /usr/local/Cellar/pkg-config/0.29.2_3 (11 files, 623.8KB) *
BREW INFO pkg-config:   Poured from bottle on 2020-05-25 at 19:03:35
BREW INFO pkg-config: From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/pkg-config.rb
BREW INFO pkg-config: License: GPL-2.0-or-later
BREW INFO pkg-config: ==> Analytics
BREW INFO pkg-config: install: 233,394 (30 days), 682,251 (90 days), 2,801,880 (365 days)
BREW INFO pkg-config: install-on-request: 31,065 (30 days), 95,053 (90 days), 578,640 (365 days)
BREW INFO pkg-config: build-error: 0 (30 days)

**Comment by doublethefish** [comment]
Update: I tried to build on a fresh(er) machine and the download worked without error and built successfuly. 👍

One difference I noticed is the machines have different openSSL cert configs in `/usr/local/etc/` which I suspect is related to the following homebrew issue:
    https://github.com/Homebrew/homebrew-core/issues/54235

The borked machine has the following extra:

-  `/usr/local/etc/openssl/**`
- `/usr/local/etc/openssl@1.1/misc/CA.pl.default`
- `/usr/local/etc/openssl@1.1/misc/tsget.pl.default`
- `/usr/local/etc/openssl@1.1/ct_log_list.cnf.dist.default`
- `/usr/local/etc/openssl@1.1/ct_log_list.cnf.default`
- `/usr/local/etc/openssl@1.1/openssl.cnf.dist.default`
- `/usr/local/etc/openssl@1.1/openssl.cnf.default`
- `/usr/local/etc/openssl@1.1/openssl/misc/CA.pl.default`

I assume that `certifi` ensures the correct paths and certs are used for the download.

I wonder how many of the other people who had this issue on the forums and stackoverflow had the same misconfiguration and would be helped by this patch.

**Comment by doublethefish** [comment]
Some more info:

- python 3.9.1 (via pyenv) downloads over https aok on origin/master (2cb084c) with the extra openSSL files mentioned above
- python 3.9.2 (via brew) downloads over https aok on 2cb084c
- the version of python 3.8.3 build that the build fails with came from python.org official download
    - IIRC there were some openSSL issues on OSX with that version of python
    - the pyenv version of 3.8.3 doesn't build on the borked machine
- python 3.8.4 (via pyenv) also downloads over https on 2cb084c with the extra openSSL files mentioned above

**Comment by misl6** [comment]
Just switched on a clean Big Sur install, and I was able to reproduce this issue.

It's not a `kivy-ios` fault. The python version downloadable via python.org specifically requires some actions in order to have a fully working certificate validation.

From their `ReadMe.rtf` file ( version 3.9.2, but should be the same for 3.8.3):
```
Certificate verification and OpenSSL

This package includes its own private copy of OpenSSL 1.1.1.   The trust certificates in system and user keychains managed by the Keychain Access application and the security command line utility are not used as defaults by the Python ssl module.  A sample command script is included in /Applications/Python 3.9 to install a curated bundle of default root certificates from the third-party certifi package (https://pypi.org/project/certifi/).  Double-click on Install Certificates to run it.

The bundled pip has its own default certificate store for verifying download connections.
```

Can you confirm that by following these instructions your issue got fixed?

**Comment by doublethefish** [comment]
Good spot/find. Seems like a good alternative workaround.

- With raw 3.9.2 python.org install (inside a virtualenv):
    - Failed.
- After running the `Install Certificates.command` script (inside the same virtualenv):
    -  Works - at least, it appears to be building and has progressed past the download section.

So, to summarize, to get kivvy to build with python versions from python.org one has to run the certs-installer script. One also needs to know that that is the case.

I guess the remaining question is "is that barrier to entry small enough for new users?". I would say "no" and still advocate for this patch, as it gets kivvy more likely to build out-of-the-box, but feel free to close if you feel it is not appropriate.

File: kivy_ios/toolchain.py [hunk 1]
```diff
@@ -85,6 +87,16 @@ def remove_junk(d):
                 unlink(join(root, fn))
 
 
+def https_download(url, filepath):
+    http = urllib3.PoolManager(cert_reqs="CERT_REQUIRED", ca_certs=certifi.where())
+    with http.request("GET", url, preload_content=False) as resp, open(
```
Comment (misl6) [inline kivy_ios/toolchain.py:92]: That call doesn't check about errors.

Example:
If a 404 is returned, it just downloads the 404 page and save that into the provided path.

File: kivy_ios/toolchain.py [hunk 2]
```diff
@@ -469,16 +481,20 @@ def report_hook(index, blksize, size):
...
-            except OSError:
-                attempts += 1
-                if attempts >= 5:
-                    logger.error('Max download attempts reached: {}'.format(attempts))
-                    raise
-                logger.warning('Download failed. Retrying in 1 second...')
-                time.sleep(1)
-                continue
+            if url.startswith("https://") or url.startswith("http://"):
+                https_download(url, filename)
```
Comment (misl6) [inline kivy_ios/toolchain.py:485]: Here you're skipping our `try: / except:` block

**Comment by misl6** [comment]
I added some comments about your changes to explain why it's a bad idea.

It's a Python (downloaded from python.org) prerequisite, not a kivy-ios one and it's specifically declared into their `Readme.rtf` file.

Maybe it's definitely a better choice to just add a clear and concise error that reminds the user about installing the certificates?

**Comment by doublethefish** [comment]
> I added some comments about your changes to explain why it's a bad idea.
You're just left feedback about the patch, valid comments that could have been fixed. That feedback doesn't invalidate the underlying concept.

I don't think I will waste further time on this project. Good luck.

**Pull request closed without merging**